### ISSUE TITLE
Point wanders sometimes when using Company Mode (Multi-Backend)

### ISSUE THREAD

**Issue by countvajhula** [greghendershott/racket-mode#592]
## Description

Sometimes, when a list of company completions is visible, point wanders to a random location in the buffer.

## Steps to reproduce

1. Create a buffer with these contents:

```
#lang racket

(define blahblah 'hello)
```

2. Save the buffer and `racket-run` to get into a REPL
3. Start typing `bla`
4. You should see more than one completion candidate. Wait a couple of seconds (not sure if this is needed but...)
5. Continue typing `blahb`
6. Once the completion candidate is unique, wait a couple of seconds
7. Point wanders to somewhere else in the REPL buffer

If it doesn't happen at first, you might need to enter at least one line into the REPL and get a result so that you are on a second or later prompt. It may also happen even when the completion candidate isn't unique.

I've noticed that this happens in regular Racket buffers too and not just the REPL buffer, but I've found the above to reliably reproduce it while formerly it was elusive.

It doesn't happen when company-mode is disabled.

In cases where it happens in a Racket source buffer, as far as I could tell, it doesn't happen when racket-xp-mode is disabled.

## Notes

This started happening within the last week and appears to be connected to the new multi-backend branch. I tried with the `master` branch and this issue is not reproducible there. Another point of difference: on master, it says "Getting information from back end about blahblah ..." next to the completion candidates. This message isn't displayed on the `rebased-multi-backend` branch.

<details>
<pre>
((alist-get 'racket-mode package-alist))
((emacs-version "27.1")
 (system-type darwin)
 (x-gtk-use-system-tooltips UNDEFINED)
 (major-mode racket-mode)
 (racket--el-source-dir "/Users/siddhartha/.emacs.d/straight/build/racket-mode/")
 (racket--rkt-source-dir "/Users/siddhartha/.emacs.d/straight/build/racket-mode/racket/")
 (racket-program "racket")
 (racket-command-timeout 10)
 (racket-path-from-emacs-to-racket-function UNDEFINED)
 (racket-path-from-racket-to-emacs-function UNDEFINED)
 (racket-browse-url-function racket-browse-url-using-temporary-file)
 (racket-documentation-search-location "https://docs.racket-lang.org/search/index.html?q=%s")
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-mode-lighter
  (:eval
   (racket--xp-mode-lighter)))
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-submodules-to-run
  ((test)
   (main)))
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-repl-history-directory "~/.emacs.d/racket-mode/")
 (racket-history-filter-regexp "\\`\\s *\\'")
 (racket-images-inline t)
 (racket-imagemagick-props nil)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "open")
 (racket-pretty-print t)
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-pseudo-tooltip)))
(enabled-minor-modes
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (buffer-ring-mode)
 (column-number-mode)
 (company-box-mode)
 (company-mode)
 (company-prescient-mode)
 (counsel-mode)
 (display-line-numbers-mode)
 (electric-indent-mode)
 (evil-collection-unimpaired-mode)
 (evil-commentary-mode)
 (evil-goggles-mode)
 (evil-local-mode)
 (evil-matchit-mode)
 (evil-mc-mode)
 (evil-mode)
 (evil-surround-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (general-override-mode)
 (global-auto-revert-mode)
 (global-company-mode)
 (global-display-line-numbers-mode)
 (global-eldoc-mode)
 (global-evil-collection-unimpaired-mode)
 (global-evil-matchit-mode)
 (global-evil-mc-mode)
 (global-evil-surround-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-tree-sitter-mode)
 (global-undo-tree-mode)
 (global-visual-line-mode)
 (global-whitespace-mode)
 (hs-minor-mode)
 (ivy-mode)
 (ivy-prescient-mode)
 (ivy-rich-mode)
 (line-number-mode)
 (mac-mouse-wheel-mode)
 (magit-auto-revert-mode)
 (menu-bar-mode)
 (override-global-mode)
 (prescient-persist-mode)
 (projectile-mode)
 (racket-xp-mode)
 (recentf-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (straight-package-neutering-mode)
 (straight-use-package-mode)
 (symex-editing-mode)
 (symex-mode)
 (telephone-line-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (undo-tree-mode)
 (visual-line-mode)
 (winner-mode)
 (yas-global-mode)
 (yas-minor-mode))
(disabled-minor-modes
 (abbrev-mode)
 (ace-window-display-mode)
 (ace-window-mode)
 (adaptive-wrap-prefix-mode)
 (all-the-icons-dired-mode)
 (archive-subfile-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (avy-linum-mode)
 (beacon-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (caps-lock-mode)
 (centaur-tabs-local-mode)
 (centaur-tabs-mode)
 (cider--debug-mode)
 (cider-auto-test-mode)
 (cider-enlighten-mode)
 (cider-mode)
 (cider-popup-buffer-mode)
 (cl-old-struct-compat-mode)
 (company-search-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (dash-fontify-mode)
 (defining-kbd-macro)
 (delete-selection-mode)
 (desktop-save-mode)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (dired-isearch-filenames-mode)
 (dired-omit-mode)
 (edebug-mode)
 (edebug-x-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (enriched-mode)
 (erc-irccontrols-mode)
 (erc-keep-place-mode)
 (erc-move-to-prompt-mode)
 (erc-noncommands-mode)
 (erc-readonly-mode)
 (erc-scrolltobottom-mode)
 (erc-smiley-mode)
 (erc-unmorse-mode)
 (evil-cleverparens-mode)
 (evil-collection-magit-toggle-text-minor-mode)
 (flyspell-mode)
 (geiser-autodoc-mode)
 (geiser-mode)
 (geiser-smart-tab-mode)
 (general-override-local-mode)
 (git-commit-mode)
 (git-timemachine-mode)
 (global-dash-fontify-mode)
 (global-hl-line-mode)
 (global-prettify-symbols-mode)
 (global-reveal-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-whitespace-newline-mode)
 (hl-line-mode)
 (horizontal-scroll-bar-mode)
 (ibuffer-auto-mode)
 (image-minor-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (ivy-rich-project-root-cache-mode)
 (jedi-mode)
 (jit-lock-debug-mode)
 (lispy-goto-mode)
 (lispy-mode)
 (lispy-other-mode)
 (mac-auto-ascii-mode)
 (mac-auto-operator-composition-mode)
 (mac-font-panel-mode)
 (macrostep-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (mml-mode)
 (mouse-wheel-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-list-checkbox-radio-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (org-table-header-line-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (rectangle-mark-mode)
 (reveal-mode)
 (rigpa-char-mode)
 (rigpa-line-mode)
 (rigpa-word-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (show-smartparens-global-mode)
 (show-smartparens-mode)
 (size-indication-mode)
 (slime-autodoc-mode)
 (slime-edit-value-mode)
 (slime-editing-mode)
 (slime-fuzzy-target-buffer-completions-mode)
 (slime-macroexpansion-minor-mode)
 (slime-mode)
 (slime-popup-buffer-mode)
 (slime-repl-map-mode)
 (slime-repl-read-mode)
 (slime-trace-dialog-autofollow-mode)
 (slime-trace-dialog-hide-details-mode)
 (slime-trace-dialog-minor-mode)
 (smartparens-global-mode)
 (smartparens-global-strict-mode)
 (smartparens-mode)
 (smartparens-strict-mode)
 (smerge-mode)
 (sr-modeline)
 (sr-popviewer-mode)
 (sr-tabs-mode)
 (sr-term-char-minor-mode)
 (sr-term-line-minor-mode)
 (straight-live-modifications-mode)
 (straight-symlink-emulation-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (tar-subfile-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (transient-resume-mode)
 (tree-sitter-hl-mode)
 (tree-sitter-mode)
 (undo-tree-visualizer-selection-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (view-mode)
 (visible-mode)
 (which-function-mode)
 (whitespace-mode)
 (whitespace-newline-mode)
 (window-divider-mode)
 (with-editor-mode)
 (xref-etags-mode))
</pre>
</details>

**Comment by greghendershott** [comment]
I've only tried to reproduce this for a few minutes now in a REPL buffer, but so far no luck.

One key difference: Yesterday I rebased onto `master` the `rebased-multi-back-end` branch. So what I'm trying includes commit 3027de0 to fix your recently reported #590.

Maybe that's why I can't reproduce it (which would be good news). Maybe the two issues are "two sides of the same coin"?? i.e. moving point in the wrong buffer is both (a) a failure to move where desired and (b) a move where not desired.

Or maybe there is some other element I don't have, like symex-mode  (I wasn't able to reproduce #590 either).

**Comment by greghendershott** [comment]
Before I saw this issue my plan for today had been to go ahead and merge the multi back end branch.

I guess I will hold off on that, at least for awhile...

**Comment by greghendershott** [comment]
I'm not sure if this is a tangent, or maybe a key to reproducing/fixing this, but:

>  Another point of difference: on master, it says "Getting information from back end about blahblah ..." next to the completion candidates. This message isn't displayed on the rebased-multi-backend branch.

I'm surprised you see that message at all, following the steps above. IIUC that would only happen if you press <kbd>F1</kbd> or <kbd>C-h</kbd> to display a documentation buffer.

[That would cause the `:company-doc-buffer` procedure, in this case `racket--repl-company-doc-buffer`, to be called, which is what displays that message.  [On `master` this "getting information..." text was inserted into the Racket Describe buffer. On the multi back end branch instead it is shown via `header-line-format`.]]

Are you using something that eagerly/proactively shows docs, without you pressing a key, just while idling on a completion candiate? Like maybe `company-quickhelp`?

**Comment by greghendershott** [comment]
OK I can reproduce this IFF I enable `company-quickhelp-local-mode` in the REPL buffer.

I'll start digging in to figure out why.

**Comment by countvajhula** [comment]
> I'm surprised you see that message at all, following the steps above. IIUC that would only happen if you press F1 or C-h to display a documentation buffer.

I also have `company-box-mode` and `company-prescient-mode` enabled. Disabling these in turn seemed to cause the issue to go away but I couldn't pinpoint which of these was involved. Glad you were able to reproduce it with `company-quickhelp-local-mode`!

**Comment by greghendershott** [comment]
All right, here's my plan:

1. As a short-term work-around, I'm going to ask you to disable `company-quickhelp-mode` in `racket-mode` and `racket-repl-mode` buffers.
2. I'm going to go ahead and merge the multi back end branch. There were already some less-obvious problems wrt quickhelp on `master`; the multi back end changes are mostly incidental IMHO.
3. I'll start a new branch to refactor some things (long story below) to fix/close this issue, for real.

# Long Story

As you may know, `company-mode` has a `:company-doc-buffer` property a mode can set to a function that takes a string and returns a buffer with help to display. Normally this buffer is shown only after the user presses a key, and it is shown in a dedicated window.

However `company-quickhelp-mode` proactively gets the string (as soon as you idle on any candidate in the list), and shows it in a `pos-tip` popup.

As for Racket Mode: Once upon a time `racket-describe` would fetch only the docs about the term in question. It was kind of a "peephole" into the full documentation page. This was ideal for `:company-doc-buffer`, so I shared a lot of code.

Over time, I found the "peephole" less-ideal for dedicated doc viewing. Racket documentation has many links, plus often you want to scroll around a page with various related items. So I moved `racket-describe-mode` to acting more like a full-fledged web browser.

Furthermore, the multi back end branch encouraged me to make it even more "async": The buffer can be shown while the contents are still being fetched, and eventually they appear. This is smoother when fetching docs from a remote host.

However, `company-quickhelp-mode` works by assuming the buffer already has the contents, and grabbing a string to put in the `pos-tip`. So for awhile now, it's been finding an empty buffer, and showing nothing (but still causing a lot of work to be done in the background). Furthermore, the fix for #590 actually contributed to this new issue, by moving point in the wrong buffer.

And, even if it did all work correctly, a `pos-tip` really wants a small amount of text; you can't scroll or click links or do anything. An entire Racket doc page is... not great.

TL;DR: Although it's good to share code between `racket-describe-mode` and `:company-doc-buffer`, probably I need to untangle this a bit. Probably the latter should even revert fully to the old "peephole" approach. And definitely it should not be async, it should wait (until some short timeout) for the doc string to be ready (or not). Then it should work more smoothly for things like `company-quickhelp-mode`.

**Comment by countvajhula** [comment]
Appreciate the detailed explanation! I had noticed the transition from the "peephole" to the full docs view, and it's nice to see your design evolution there. I don't mind disabling company-box mode (or whichever one is responsible for the proactive doc-lookup).

**Comment by xgqt** [comment]
I have the same issue & can confirm it seems to be caused by company-quickhelp.

**Comment by greghendershott** [comment]
Don't worry, I'm still working on this.

I got side-tracked/speed-bumped until I realized my installed version of `company-quickhelp-mode` was `fill-paragraph`-ing things to fit in the pos-tip, munging the nice layout of some Racket docs.

Having just digested that, I'm in a "measure twice, cut once" mindset. But I do want to merge something in the next day or two.

**Comment by greghendershott** [comment]
OK, I believe I now have something that works well for the default `company-mode` case (where it shows a normal buffer in a window), as well as the `company-quickhelp-mode` case (I verified that more-recent versions of that preserve formatting).

So I plan to squash and merge those commits soon-ish.

**Comment by greghendershott** [comment]
Merged. Please let me know if you have any questions or still experience any problems.

**Comment by xgqt** [comment]
Work fine for me. Thanks a lot!

**Comment by countvajhula** [comment]
This is so nifty that I kinda want to be able to pull it up on demand even outside a completion context. Anyhow, it's great, thank you!

**Comment by greghendershott** [comment]
You're welcome.

It is pretty nifty!

However I think using `racket-describe` is more-reliably nifty. :smile:

What I mean: One reason I eventually put so much work into changing `racket-describe` into more of a "web browser" experience is because so much Racket documentation is written with that expectation/assumption:

- Often it has links you need/want to follow.

- The "peephole" approach fails for docs written assuming you're able to scroll up/down around the immediate topic, for context.

In a completion context, the transient popup style is kind of what you have to live with, so you have to accept those limitations. And I'm glad I revived that, with this commit. And for some topics, the limits don't even matter; sometimes you get lucky. But for other topics, although it initially looks nifty... it turns out to be an impoverished experience.

Anyway that's my personal take on it.

**Comment by countvajhula** [comment]
Yeah, I agree with that. Personally, I find myself most wishing for a slightly different configuration of these, something like:

1. a simple signature suggestion.
2. a lightweight summary doc view, a lot like the "peephole" you added, but arguably trimmed down even more to (say) just the first paragraph of docs.
3. a full doc mode (i.e. `racket-describe`) for greater context and navigation rather than just lookup
4. an escape to the browser docs.

... In decreasing order of lightweight-ness, and I think I would probably use them in decreasing order of frequency as well, while using all of them often.

I think my preference would be for (1) to always be shown (to avoid the doc lookup altogether for this simplest case), and for (2), (3) and (4) to be easily brought up on demand. And maybe (2) and (3) could even be part of the same usage flow -- e.g. `C-c C-d` first pulls up (2), and then hitting `C-d` again goes into `racket-describe` mode. The peephole being trimmed down also addresses some of the issues you pointed out, and narrows the scope of this feature while encouraging (3) for anything more than the simplest lookups. Oh, and, these are not suggestions so much as thoughts thunk out loud 😝 

FWIW, similar experiences for (1) in other IDEs are, in DrRacket, there's an arrow at the top right of the screen which if you mouse over (or even leave your mouse over), it shows you the signature of the form you are engaged in writing down. For ELisp and also python-mode, Emacs shows the signature in the mode line and I find these to be helpful.
